// @camunda8/sdk/dist/c8/lib/C8Dto.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GetVariableResponse = exports.SearchUsersResponse = exports.EvaluateDecisionResponse = exports.EvaluatedDecision = exports.EvaluatedInput = exports.MatchedRule = exports.EvaluatedOutput = exports.UploadDocumentsResponse = exports.UploadDocumentResponse = exports.PublishMessageResponse = exports.CorrelateMessageResponse = exports.BroadcastSignalResponse = exports.CreateProcessInstanceWithResultResponse = exports.CreateProcessInstanceResponse = exports.DeployResourceResponse = exports.FormDeployment = exports.DecisionRequirementsDeployment = exports.DecisionDeployment = exports.ProcessDeployment = exports.RestApiJob = void 0;
const lib_1 = require("../../lib");
const C8DtoInternal_1 = require("./C8DtoInternal");
class RestApiJob extends lib_1.LosslessDto {
}
exports.RestApiJob = RestApiJob;
__decorate([
    lib_1.Int64String
], RestApiJob.prototype, "jobKey", void 0);
__decorate([
    lib_1.Int64String
], RestApiJob.prototype, "processInstanceKey", void 0);
__decorate([
    lib_1.Int64String
], RestApiJob.prototype, "processDefinitionKey", void 0);
__decorate([
    lib_1.Int64String
], RestApiJob.prototype, "elementInstanceKey", void 0);
__decorate([
    lib_1.Int64String
], RestApiJob.prototype, "deadline", void 0);
class ProcessDeployment extends lib_1.LosslessDto {
}
exports.ProcessDeployment = ProcessDeployment;
__decorate([
    lib_1.Int64String
], ProcessDeployment.prototype, "processDefinitionKey", void 0);
class DecisionDeployment extends lib_1.LosslessDto {
}
exports.DecisionDeployment = DecisionDeployment;
__decorate([
    lib_1.Int64String
], DecisionDeployment.prototype, "decisionRequirementsKey", void 0);
class DecisionRequirementsDeployment extends lib_1.LosslessDto {
}
exports.DecisionRequirementsDeployment = DecisionRequirementsDeployment;
__decorate([
    lib_1.Int64String
], DecisionRequirementsDeployment.prototype, "decisionRequirementsKey", void 0);
class FormDeployment {
}
exports.FormDeployment = FormDeployment;
__decorate([
    lib_1.Int64String
], FormDeployment.prototype, "formKey", void 0);
class DeployResourceResponse extends C8DtoInternal_1.DeployResourceResponseDto {
}
exports.DeployResourceResponse = DeployResourceResponse;
class CreateProcessInstanceResponse {
}
exports.CreateProcessInstanceResponse = CreateProcessInstanceResponse;
__decorate([
    lib_1.Int64String
], CreateProcessInstanceResponse.prototype, "processDefinitionKey", void 0);
__decorate([
    lib_1.Int64String
], CreateProcessInstanceResponse.prototype, "processInstanceKey", void 0);
class CreateProcessInstanceWithResultResponse extends CreateProcessInstanceResponse {
}
exports.CreateProcessInstanceWithResultResponse = CreateProcessInstanceWithResultResponse;
/** The signal was broadcast. */
class BroadcastSignalResponse extends lib_1.LosslessDto {
}
exports.BroadcastSignalResponse = BroadcastSignalResponse;
__decorate([
    lib_1.Int64String
    /** The unique ID of the signal that was broadcast. */
], BroadcastSignalResponse.prototype, "signalKey", void 0);
class CorrelateMessageResponse extends lib_1.LosslessDto {
}
exports.CorrelateMessageResponse = CorrelateMessageResponse;
__decorate([
    lib_1.Int64String
], CorrelateMessageResponse.prototype, "messageKey", void 0);
__decorate([
    lib_1.Int64String
], CorrelateMessageResponse.prototype, "processInstanceKey", void 0);
class PublishMessageResponse extends lib_1.LosslessDto {
}
exports.PublishMessageResponse = PublishMessageResponse;
__decorate([
    lib_1.Int64String
], PublishMessageResponse.prototype, "messageKey", void 0);
class UploadDocumentResponse extends lib_1.LosslessDto {
    constructor() {
        super(...arguments);
        /** Document discriminator. Always set to "camunda". */
        this['camunda.document.type'] = 'camunda';
    }
}
exports.UploadDocumentResponse = UploadDocumentResponse;
class UploadDocumentsResponse {
}
exports.UploadDocumentsResponse = UploadDocumentsResponse;
class EvaluatedOutput extends lib_1.LosslessDto {
}
exports.EvaluatedOutput = EvaluatedOutput;
class MatchedRule extends lib_1.LosslessDto {
}
exports.MatchedRule = MatchedRule;
__decorate([
    (0, lib_1.ChildDto)(EvaluatedOutput)
], MatchedRule.prototype, "evaluatedOutputs", void 0);
class EvaluatedInput extends lib_1.LosslessDto {
}
exports.EvaluatedInput = EvaluatedInput;
class EvaluatedDecision extends lib_1.LosslessDto {
}
exports.EvaluatedDecision = EvaluatedDecision;
__decorate([
    (0, lib_1.ChildDto)(MatchedRule)
], EvaluatedDecision.prototype, "matchedRules", void 0);
__decorate([
    (0, lib_1.ChildDto)(EvaluatedInput)
], EvaluatedDecision.prototype, "evaluatedInputs", void 0);
class EvaluateDecisionResponse extends lib_1.LosslessDto {
}
exports.EvaluateDecisionResponse = EvaluateDecisionResponse;
__decorate([
    (0, lib_1.ChildDto)(EvaluatedDecision)
], EvaluateDecisionResponse.prototype, "evaluatedDecisions", void 0);
class UserItem extends lib_1.LosslessDto {
}
__decorate([
    lib_1.Int64String
], UserItem.prototype, "id", void 0);
/** The user search result. */
class SearchUsersResponse extends lib_1.LosslessDto {
}
exports.SearchUsersResponse = SearchUsersResponse;
__decorate([
    (0, lib_1.ChildDto)(UserItem)
], SearchUsersResponse.prototype, "items", void 0);
class GetVariableResponse {
}
exports.GetVariableResponse = GetVariableResponse;
// export interface DecisionInstanceSearchFilter {
// 	/** System-generated key for a decision evaluation instance. */
// 	decisionEvaluationInstanceKey?: string | AdvancedStringFilter
// 	/** The state of the decision instance. */
// 	state?: 'EVALUATED' | 'FAILED' | 'UNKNOWN' | 'UNSPECIFIED'
// 	/* The evaluation failure of the decision instance. */
// 	evaluationFailure?: string
// 	/** The evaluation date. */
// 	evaluationDate?: string | AdvancedDateTimeFilter
// 	/** The decision definition ID. */
// 	decisionDefinitionId?: string | AdvancedStringFilter
// 	/** The decision definition key. */
// 	decisionDefinitionKey?: string | AdvancedStringFilter
// 	/** The decision definition name. */
// 	decisionDefinitionName?: string | AdvancedStringFilter
// 	/** The decision definition version. */
// 	decisionDefinitionVersion?: number | AdvancedNumberFilter
// 	/** The process definition key associated to this decision instance. */
// 	processDefinitionKey?: string | AdvancedStringFilter
// 	/** The process instance key associated to this decision instance. */
// 	processInstanceKey?: string | AdvancedStringFilter
// 	/** The key of the parent decision evaluation. Note that this is not the identifier of an individual decision instance; the decisionEvaluationInstanceKey is the identifier for a decision instance. */
// 	decisionEvaluationKey?: string
// 	/** The key of the element instance this decision instance is linked to. */
// 	elementInstanceKey?: string | AdvancedStringFilter
// 	/** The tenant ID. */
// 	tenantId?: string | AdvancedStringFilter
// 	/** The decision type. */
// 	decisionDefinitionType?:
// 		| 'DECISION_TABLE'
// 		| 'LITERAL_EXPRESSION'
// 		| 'UNSPECIFIED'
// 		| 'UNKNOWN'
// }
// export interface SearchDecisionInstancesRequest
// 	extends BaseSearchRequest<
// 		| 'decisionDefinitionId'
// 		| 'decisionDefinitionKey'
// 		| 'decisionDefinitionName'
// 		| 'decisionDefinitionType'
// 		| 'decisionDefinitionVersion'
// 		| 'decisionEvaluationInstanceKey'
// 		| 'decisionEvaluationKey'
// 		| 'elementInstanceKey'
// 		| 'evaluationDate'
// 		| 'evaluationFailure'
// 		| 'processDefinitionKey'
// 		| 'processInstanceKey'
// 		| 'state'
// 		| 'tenantId',
// 		DecisionInstanceSearchFilter
// 	> {}
// interface DecisionInstanceDetails {
// 	/** The decision evaluation instance key. */
// 	decisionEvaluationInstanceKey: string
// 	/** The decision definition ID. */
// 	decisionDefinitionId: string
// 	/** The decision definition key. */
// 	decisionDefinitionKey: string
// 	/** The decision definition name. */
// 	decisionDefinitionName: string
// 	/** The decision definition version. */
// 	decisionDefinitionVersion: number
// 	/** The process definition key associated to this decision instance. */
// 	processDefinitionKey: string
// 	/** The process instance key associated to this decision instance. */
// 	processInstanceKey: string
// 	/** The state of the decision instance. */
// 	state: 'EVALUATED' | 'FAILED' | 'UNKNOWN' | 'UNSPECIFIED'
// 	/** The evaluation date. */
// 	evaluationDate: string
// 	/** The evaluation failure message, if any. */
// 	evaluationFailure?: string
// 	/** The tenant ID. */
// 	tenantId: string
// 	/** The decision type. */
// 	decisionDefinitionType:
// 		| 'DECISION_TABLE'
// 		| 'LITERAL_EXPRESSION'
// 		| 'UNSPECIFIED'
// 		| 'UNKNOWN'
// 	/** The result of the decision evaluation. */
// 	result: string
// 	/** The ID of the decision instance. */
// 	decisionInstanceId: string
// }
// export interface SearchDecisionInstancesResponse
// 	extends PaginatedSearchResponse<DecisionInstanceDetails> {}
/**
 * Response from getting a single decision instance by its key.
 */
// export interface GetDecisionInstanceResponse {
// 	/** The decision instance key. Note that this is not the unique identifier of the entity itself; the decisionInstanceId serves as the primary identifier. */
// 	decisionInstanceKey: string
// 	/** The decision definition ID. */
// 	decisionDefinitionId: string
// 	/** The decision definition key. */
// 	decisionDefinitionKey: string
// 	/** The decision definition name. */
// 	decisionDefinitionName: string
// 	/** The decision definition version. */
// 	decisionDefinitionVersion: number
// 	/** The process definition key associated to this decision instance. */
// 	processDefinitionKey: string
// 	/** The process instance key associated to this decision instance. */
// 	processInstanceKey: string
// 	/** The state of the decision instance. */
// 	state: 'EVALUATED' | 'FAILED' | 'UNKNOWN' | 'UNSPECIFIED'
// 	/** The evaluation date. */
// 	evaluationDate: string
// 	/** The evaluation failure message, if any. */
// 	evaluationFailure?: string
// 	/** The tenant ID. */
// 	tenantId: string
// 	/** The decision type. */
// 	decisionDefinitionType:
// 		| 'DECISION_TABLE'
// 		| 'LITERAL_EXPRESSION'
// 		| 'UNSPECIFIED'
// 		| 'UNKNOWN'
// 	/** The result of the decision evaluation. */
// 	result: string
// 	/** The evaluated inputs of the decision instance. */
// 	evaluatedInputs: Array<{
// 		/** The ID of the evaluated decision input. */
// 		inputId: string
// 		/** The name of the evaluated decision input. */
// 		inputName: string
// 		/** The value of the evaluated decision input. */
// 		inputValue: string
// 	}>
// 	matchedRules: Array<{
// 		/** The ID of the matched rule. */
// 		ruleId: string
// 		/** The index of the matched rule. */
// 		ruleIndex: number
// 		evaluatedOutputs: Array<{
// 			/** The ID of the evaluated decision output. */
// 			outputId: string
// 			/** The name of the evaluated decision output. */
// 			outputName: string
// 			/** The value of the evaluated decision output. */
// 			outputValue: string
// 		}>
// 	}>
// }
